src/node/mini_miner.cpp

```diff
@@ -0,0 +1,318 @@
+                for (const auto& outpoint : outpoints_it->second) {
+                    m_bump_fees.emplace(outpoint, 0);
+                }
+                m_requested_outpoints_by_txid.erase(outpoints_it);
+            }
+        }
+    }
+
+    // Build the m_descendant_set_by_txid cache.
+    for (const auto& txiter : cluster) {
+        const auto& txid = txiter->GetTx().GetHash();
+        // Cache descendants for future use. Unlike the real mempool, a descendant MiniMinerMempoolEntry
+        // will not exist without its ancestor MiniMinerMempoolEntry, so these sets won't be invalidated.
+        std::vector<MockEntryMap::iterator> cached_descendants;
+        const bool remove{m_to_be_replaced.count(txid) > 0};
+        CTxMemPool::setEntries descendants;
+        mempool.CalculateDescendants(txiter, descendants);
+        Assume(descendants.count(txiter) > 0);
+        for (const auto& desc_txiter : descendants) {
+            const auto txid_desc = desc_txiter->GetTx().GetHash();
+            const bool remove_desc{m_to_be_replaced.count(txid_desc) > 0};
+            auto desc_it{m_entries_by_txid.find(txid_desc)};
+            Assume((desc_it == m_entries_by_txid.end()) == remove_desc);
+            if (remove) Assume(remove_desc);
+            // It's possible that remove=false but remove_desc=true.
+            if (!remove && !remove_desc) {
+                cached_descendants.push_back(desc_it);
+            }
+        }
+        if (remove) {
+            Assume(cached_descendants.empty());
+        } else {
+            m_descendant_set_by_txid.emplace(txid, cached_descendants);
+        }
+    }
+
+    // Release the mempool lock; we now have all the information we need for a subset of the entries
+    // we care about. We will solely operate on the MiniMinerMempoolEntry map from now on.
+    Assume(m_in_block.empty());
+    Assume(m_requested_outpoints_by_txid.size() <= outpoints.size());
+    SanityCheck();
+}
+
+// Compare by min(ancestor feerate, individual feerate), then iterator
+struct AncestorFeerateComparator
+{
+    template<typename I>
+    bool operator()(const I& a, const I& b) const {
+        const auto& ea = a->second;
+        const CFeeRate a_feerate =
+            ea.GetModFeesWithAncestors() * ea.GetTxSize() <= ea.GetModifiedFee() * ea.GetSizeWithAncestors() ?
+            CFeeRate(ea.GetModFeesWithAncestors(), ea.GetSizeWithAncestors()) :
+            CFeeRate(ea.GetModifiedFee(), ea.GetTxSize());
+        const auto& eb = b->second;
+        const CFeeRate b_feerate =
+            eb.GetModFeesWithAncestors() * eb.GetTxSize() <= eb.GetModifiedFee() * eb.GetSizeWithAncestors() ?
+            CFeeRate(eb.GetModFeesWithAncestors(), eb.GetSizeWithAncestors()) :
+            CFeeRate(eb.GetModifiedFee(), eb.GetTxSize());
+        if (a_feerate != b_feerate) {
+            return a_feerate > b_feerate;
+        }
+        // Compare by txid
+        return a->first < b->first;
+    }
+};
```

Comment:
This may be easier to comprehend:
```
// Compare by min(ancestor feerate, individual feerate), then iterator.
//
// The ancestor feerate of a high-feerate tx should be reduced by its
// low-feerate ancestors, but the ancestor feerate of a low-feerate tx
// should not be increased by its high-feerate ancestors.
struct AncestorFeerateComparator
{
    template<typename I>
    bool operator()(const I& a, const I& b) const {
        auto min_feerate = [](const MiniMinerMempoolEntry& e) -> CFeeRate {
            const CAmount afee{e.GetModFeesWithAncestors()};
            const int64_t asize{e.GetSizeWithAncestors()};
            const CAmount fee{e.GetModifiedFee()};
            const int64_t size{e.GetTxSize()};
            // Comparing ancestor feerate with individual feerate:
            //     afee / asize <= fee / size
            // Avoid division and possible loss of precision by
            // multiplying both sides by the sizes:
            return afee * size < fee * asize ?
                       CFeeRate(afee, asize) :
                       CFeeRate(fee, size);
        };
        CFeeRate a_feerate{min_feerate(a->second)};
        CFeeRate b_feerate{min_feerate(b->second)};
        if (a_feerate != b_feerate) {
            return a_feerate > b_feerate;
        }
        // Compare by txid
        return a->first < b->first;
    }
};
```

Reply:
Thanks I took your suggestion with slightly more elaborate variable names.